libs/api/domains/powerbi/src/lib/powerbi.resolver.ts

```diff
@@ -0,0 +1,176 @@
+import { Configuration, ConfidentialClientApplication } from '@azure/msal-node'
+import { Inject } from '@nestjs/common'
+import { Args, Query, Resolver } from '@nestjs/graphql'
+import type { Logger } from '@island.is/logging'
+import { LOGGER_PROVIDER } from '@island.is/logging'
+import { PowerBiEmbedTokenInput } from './dto/powerbiEmbedToken.input'
+import { PowerBiEmbedTokenResponse } from './models/powerbiEmbedTokenResponse'
+
+type Owner = 'Fiskistofa'
+const BASE_URL = 'https://api.powerbi.com/v1.0/myorg'
+
+@Resolver()
+export class PowerBiResolver {
+  constructor(
+    @Inject(LOGGER_PROVIDER)
+    private readonly logger: Logger,
+  ) {}
+
+  private async getAccessToken(owner: Owner) {
+    let clientId = ''
+    let clientSecret = ''
+    let tenantId = ''
+
+    switch (owner) {
+      case 'Fiskistofa':
+        {
+          clientId = process.env.FISKISTOFA_POWERBI_CLIENT_ID as string
+          clientSecret = process.env.FISKISTOFA_POWERBI_CLIENT_SECRET as string
+          tenantId = process.env.FISKISTOFA_POWERBI_TENANT_ID as string
+        }
+        break
+      default: {
+        this.logger.warn(
+          'User requested Power Bi embed token without providing an owner',
+        )
+        return null
+      }
+    }
+
+    const msalConfig: Configuration = {
+      auth: {
+        clientId,
+        clientSecret,
+        authority: `https://login.microsoftonline.com/${tenantId}`,
+      },
+    }
+
+    const clientApplication = new ConfidentialClientApplication(msalConfig)
+
+    try {
+      const tokenResponse = await clientApplication.acquireTokenByClientCredential(
+        {
+          scopes: ['https://analysis.windows.net/powerbi/api/.default'],
+        },
+      )
+      return tokenResponse?.accessToken
+    } catch (err) {
+      this.logger.error(
+        'Error occurred when trying to fetch Power Bi accessToken',
+        err,
+      )
+      return null
+    }
+  }
+
+  private async getReport(
+    workspaceId: string,
+    reportId: string,
+    accessToken: string,
+  ) {
+    const url = `${BASE_URL}/groups/${workspaceId}/reports/${reportId}`
+    try {
+      const response = await fetch(url, {
+        method: 'GET',
+        headers: {
+          'Content-Type': 'application/json',
+          Authorization: `Bearer ${accessToken}`,
+        },
+      })
+      if (!response.ok) {
+        this.logger.error('Power Bi report response was not 2xx')
+        return null
+      }
+      const report = await response.json()
+      return report
+    } catch (err) {
+      this.logger.error(`Error occurred when trying to fetch Power Bi report`, {
+        workspaceId,
+        reportId,
+        err,
+      })
+      return null
```

Comment:
Not recommended to do custom error handling and swallow all errors. Better to use standard error handling and handle errors as close to the UI as possible.

You may differentiate client errors from server errors. For example ok to return null if some "id" doesn't exist, but unexpected errors (missing config, 500 responses, network errors) should be thrown and handled higher up.

Enhanced Fetch can help with this and includes consistent error logging.

Reply:
Ok, that's something I've been wondering about. How we think about handling errors.

I'll change it